src/server/scripts/Kalimdor/CavernsOfTime/CullingOfStratholme/boss_mal_ganis.cpp

```diff
@@ -30,219 +30,179 @@ Script Data End */
-                        bYelled = true;
-                    }
-
-                    if (!bYelled2 && HealthBelowPct(15))
-                    {
-                        Talk(SAY_15HEALTH);
-                        bYelled2 = true;
-                    }
+                if (!_defeated)
+                    instance->SetBossState(DATA_MAL_GANIS, NOT_STARTED);
+            }
 
-                    if (HealthBelowPct(1))
-                    {
-                        //Handle Escape Event: Don't forget to add Player::RewardPlayerAndGroupAtEvent
-                        me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_NOT_SELECTABLE);
-                        uiOutroStep = 1;
-                        Phase = OUTRO;
+            void DamageTaken(Unit* source, uint32 &damage) override
+            {
+                if (damage >= me->GetHealth())
+                {
+                    damage = me->GetHealth()-1;
+                    if (_defeated)
                         return;
-                    }
-
-                    if (Creature* pArthas = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_ARTHAS)))
-                        if (pArthas->isDead())
+                    _defeated = true;
+                    if (InstanceMap* map = instance->instance->ToInstanceMap())
+                        if (map->IsHeroic())
                         {
-                            EnterEvadeMode();
-                            me->DisappearAndDie();
-                            instance->SetBossState(DATA_MAL_GANIS, FAIL);
+                            Player* sourcePlayer = source->GetCharmerOrOwnerPlayerOrPlayerItself();
+                            if (!sourcePlayer)
+                            {
+                                MapRefManager const& players = map->GetPlayers();
+                                // Find any player with a valid instance state
+                                for (MapRefManager::const_iterator it = players.begin(); !sourcePlayer && it != players.end(); ++it)
+                                {
+                                    Player* p = it->GetSource();
+                                    if (p && !p->IsGameMaster() && p->m_InstanceValid)
+                                        sourcePlayer = p;
+                                }
+                                // Fall back to any player, regardless of instance state
+                                for (MapRefManager::const_iterator it = players.begin(); !sourcePlayer && it != players.end(); ++it)
+                                {
+                                    Player* p = it->GetSource();
+                                    if (p && !p->IsGameMaster())
+                                        sourcePlayer = p;
+                                }
+                            }
+
+                            if (sourcePlayer)
+                                map->PermBindAllPlayers(sourcePlayer);
```

Comment:
TBF I'm pretty sure perm bind logic is broken anyway. It should bind all players in zone regardless of group state, opens up all kinds of exploits probably (dropping group on chest bosses before dies, don't get saved, then regroup and loot chest etc).